core/src/net/tcp_stream.cpp

```diff
@@ -0,0 +1,165 @@
+    {
+        CUBOS_ERROR("Failed to connect TCP stream to {}:{}: {}", address.toString(), port, logSystemError());
+        return false;
+    }
+
+    CUBOS_DEBUG("New TCP stream at address {} and port '{}'", address.toString(), port);
+    return true;
+}
+
+void TcpStream::inner(InnerSocket socket)
+{
+    disconnect();
+    mSock = socket;
+    CUBOS_DEBUG("New TCP stream at socket {}", mSock);
+}
+
+void TcpStream::disconnect()
+{
+    if (mSock != InnerInvalidSocket)
+    {
+        CUBOS_WARN("Closing TCP stream socket: '{}'", mSock);
+#ifdef _WIN32
+        closesocket(mSock);
+        WSACleanup();
+#else
+        close(mSock);
+#endif
+        mSock = InnerInvalidSocket;
+    }
+}
+
+TcpStream::TcpStream()
+{
+#ifdef _WIN32
+    WSADATA wsa;
+    CUBOS_ASSERT(WSAStartup(MAKEWORD(2, 2), &wsa) == 0, " WSAStartup failed: {}", WSAGetLastError());
+#endif
+}
+
+TcpStream::~TcpStream()
+{
+    disconnect();
+}
+
+std::size_t TcpStream::read(void* buffer, std::size_t size)
+{
+    if (mSock == InnerInvalidSocket)
+    {
+        CUBOS_ERROR("Invalid socket");
+        return 0;
+    }
+
+#ifdef _WIN32
+    auto bytesRead = ::recv(mSock, static_cast<char*>(buffer), static_cast<int>(size), 0);
+#else
+    auto bytesRead = ::recv(mSock, buffer, size, 0);
+#endif
+    if (bytesRead < 0)
+    {
+        CUBOS_ERROR("Failed to read from TCP stream: {}", logSystemError());
+        return 0;
+    }
+
+    if (bytesRead == 0)
+    {
+        CUBOS_WARN("TCP stream EOF");
+        mEof = true;
+    }
+
+    CUBOS_INFO("Incoming TCP message");
+    auto bytesReadCasted = static_cast<std::size_t>(bytesRead);
+    mPos += bytesReadCasted;
+    return bytesReadCasted;
+}
+
+std::size_t TcpStream::write(const void* buffer, std::size_t size)
+{
+    if (mSock == InnerInvalidSocket)
+    {
+        CUBOS_ERROR("Invalid socket");
+        return 0;
+    }
+
+#ifdef _WIN32
+    auto bytesWritten = ::send(mSock, static_cast<const char*>(buffer), static_cast<int>(size), 0);
+#else
+    auto bytesWritten = ::send(mSock, buffer, size, 0);
+#endif
+    if (bytesWritten < 0)
+    {
+        CUBOS_ERROR("Failed to write TCP buffer: {}", logSystemError());
+        return 0;
+    }
+
+    CUBOS_INFO("Sent TCP message");
+    auto bytesWrittenCasted = static_cast<std::size_t>(bytesWritten);
+    mPos += bytesWrittenCasted;
+    return bytesWrittenCasted;
+}
+
+char TcpStream::peek() const
+{
+    if (mSock == InnerInvalidSocket)
+    {
+        CUBOS_ERROR("Invalid socket");
+        return ' ';
+    }
+
+    char c;
+
+#ifdef _WIN32
+    auto bytesRead = ::recv(mSock, &c, 1, MSG_PEEK);
+#else
+    auto bytesRead = ::recv(mSock, static_cast<void*>(&c), 1, MSG_PEEK);
+#endif
+    if (bytesRead < 0)
+    {
+        CUBOS_ERROR("Failed to peek from TCP stream: {}", logSystemError());
+        return false;
+    }
+
+    if (bytesRead == 0)
+    {
+        CUBOS_WARN("TCP stream EOF");
+        // TODO: cant change mEof to true as its a const virtual method...
```

Comment:
It definitely should not be const, whoops. Feel free to change that!
Also, this peek implementation is a bit weird, as it erases data permanently.
Peek is supposed to be something like 'see what comes next without actually removing it from the stream, or advancing the cursor'.
You can implement it by storing an extra `char buffer` field or something, and a flag of whether that has been read or not before. If the flag is set when you call `peek`, you just return the existing buffer. Otherwise you write to it and set the flag. If the flag is set when you call `read`, you first copy the buffer to the output, and then continue reading if necessary; and you also should reset the flag.